sql/core/src/test/scala/org/apache/spark/sql/PlanStabilitySuite.scala

```diff
@@ -0,0 +1,300 @@
+   *
+   * @param plan    the [[SparkPlan]]
+   * @param name    the name of the query
+   * @param variant if false, this plan will become the only approved plan, otherwise
+   *                it will be added as an additional approved plan.
+   * @param explain the full explain output; this is saved to help debug later as the simplified
+   *                plan is not too useful for debugging
+   */
+  private def approvePlan(
+      plan: SparkPlan,
+      name: String,
+      variant: Boolean,
+      explain: String): Unit = {
+    val dir = getDirForTest(name)
+    val simplified = getSimplifiedPlan(plan)
+    val foundMatch = dir.exists() && isApproved(name, dir, simplified)
+
+    if (!foundMatch) {
+      if (!variant) {
+        FileUtils.deleteDirectory(dir)
+        assert(dir.mkdirs())
+      }
+      val nextVariant = getNextVariantNumber(name)
+      val file = new File(dir, s"$nextVariant.simplified.txt")
+      FileUtils.writeStringToFile(file, simplified, StandardCharsets.UTF_8)
+      val fileOriginalPlan = new File(dir, s"$nextVariant.explain.txt")
+      FileUtils.writeStringToFile(fileOriginalPlan, explain, StandardCharsets.UTF_8)
+      logInfo(s"APPROVED: $file $fileOriginalPlan")
+    }
+  }
+
+  private def checkWithApproved(plan: SparkPlan, name: String, explain: String): Unit = {
+    val dir = getDirForTest(name)
+    val tempDir = FileUtils.getTempDirectory
+    val existingVariants = getExistingVariants(name)
+    val actualSimplified = getSimplifiedPlan(plan)
+    val foundMatch = isApproved(name, dir, actualSimplified)
+
+    if (!foundMatch) {
+      // show diff with last approved
+      val approvedSimplifiedFile = new File(dir, s"${existingVariants.max}.simplified.txt")
+      val approvedExplainFile = new File(dir, s"${existingVariants.max}.explain.txt")
+
+      val actualSimplifiedFile = new File(tempDir, s"$name.actual.simplified.txt")
+      val actualExplainFile = new File(tempDir, s"$name.actual.explain.txt")
+
+      val approvedSimplified = FileUtils.readFileToString(
+        approvedSimplifiedFile, StandardCharsets.UTF_8)
+      // write out for debugging
+      FileUtils.writeStringToFile(actualSimplifiedFile, actualSimplified, StandardCharsets.UTF_8)
+      FileUtils.writeStringToFile(actualExplainFile, explain, StandardCharsets.UTF_8)
+
+      val header =
+        s"""
+           |Plans did not match:
+           |last approved plan: ${approvedSimplifiedFile.getAbsolutePath}
+           |last explain: ${approvedExplainFile.getAbsolutePath}
+           |actual plan: ${actualSimplifiedFile.getAbsolutePath}
+           |actual explain: ${actualExplainFile.getAbsolutePath}
+        """.stripMargin
+      val msg =
+        s"""
+           |$header
+           |${sideBySide(approvedSimplified.linesIterator.toSeq,
+          actualSimplified.linesIterator.toSeq).mkString("\n")}
+         """.stripMargin
+      fail(msg)
+    }
+  }
+
+  private def getSimplifiedPlan(plan: SparkPlan): String = {
```

Comment:
Can you explain the simplifying logic here by adding some method comments?

Reply:
added: 3186b87